File: app/src/main/java/com/coinninja/coinkeeper/service/DeleteWalletService.java
package com.coinninja.coinkeeper.service;

import android.app.IntentService;
import android.content.Intent;

import com.coinninja.coinkeeper.CoinKeeperApplication;
import com.coinninja.coinkeeper.cn.service.PushNotificationDeviceManager;
import com.coinninja.coinkeeper.cn.service.PushNotificationEndpointManager;
import com.coinninja.coinkeeper.cn.wallet.SyncWalletManager;
import com.coinninja.coinkeeper.model.helpers.DaoSessionManager;
import com.coinninja.coinkeeper.model.helpers.UserHelper;
import com.coinninja.coinkeeper.service.client.SignedCoinKeeperApiClient;
import com.coinninja.coinkeeper.util.Intents;
import com.coinninja.coinkeeper.util.analytics.Analytics;
import com.coinninja.coinkeeper.util.android.LocalBroadCastUtil;

import javax.inject.Inject;

import androidx.annotation.Nullable;
import dagger.android.AndroidInjection;

public class DeleteWalletService extends IntentService {

    public static final String TAG = DeleteWalletService.class.getName();
    @Inject
    PushNotificationDeviceManager pushNotificationDeviceManager;
    @Inject
    PushNotificationEndpointManager pushNotificationEndpointManager;
    @Inject
    SignedCoinKeeperApiClient apiClient;
    @Inject
    DaoSessionManager daoSessionManager;
    @Inject
    Analytics analytics;
    @Inject
    UserHelper userHelper;
    @Inject
    LocalBroadCastUtil localBroadCastUtil;
    @Inject
    CoinKeeperApplication application;
    @Inject
    SyncWalletManager syncWalletManager;

    public DeleteWalletService() {
        this(TAG);
    }

    public DeleteWalletService(String name) {
        super(name);
    }

    @Override
    public void onCreate() {
        AndroidInjection.inject(this);
        super.onCreate();
        localBroadCastUtil = new LocalBroadCastUtil(this);
    }

    @Override
    protected void onHandleIntent(@Nullable Intent intent) {
        syncWalletManager.cancelAllScheduledSync();
        if (pushNotificationEndpointManager.hasEndpoint()) {
            pushNotificationEndpointManager.unRegister();
            pushNotificationEndpointManager.removeEndpoint();
        }
        pushNotificationDeviceManager.removeCNDevice();
        apiClient.resetWallet();
        daoSessionManager.resetAll();
        userHelper.createFirstUser();
        resetAnalyticsProperties();
        localBroadCastUtil.sendBroadcast(Intents.ACTION_ON_WALLET_DELETED);
    }

    private void resetAnalyticsProperties() {
        analytics.trackEvent(Analytics.EVENT_WALLET_DELETE);
        analytics.setUserProperty(Analytics.PROPERTY_HAS_WALLET, false);
        analytics.setUserProperty(Analytics.PROPERTY_PHONE_VERIFIED, false);
        analytics.setUserProperty(Analytics.PROPERTY_HAS_WALLET_BACKUP, false);
        analytics.setUserProperty(Analytics.PROPERTY_HAS_BTC_BALANCE, false);
        analytics.flush();
    }

}

File: app/src/main/java/com/coinninja/coinkeeper/service/DeleteWalletService.kt
package com.coinninja.coinkeeper.service

import android.app.IntentService
import android.content.Intent
import com.coinninja.coinkeeper.CoinKeeperApplication
import com.coinninja.coinkeeper.cn.service.PushNotificationDeviceManager
import com.coinninja.coinkeeper.cn.service.PushNotificationEndpointManager
import com.coinninja.coinkeeper.cn.wallet.CNWalletManager
import com.coinninja.coinkeeper.cn.wallet.SyncWalletManager
import com.coinninja.coinkeeper.service.client.SignedCoinKeeperApiClient
import com.coinninja.coinkeeper.twitter.MyTwitterProfile
import com.coinninja.coinkeeper.util.DropbitIntents
import com.coinninja.coinkeeper.util.analytics.Analytics
import com.coinninja.coinkeeper.util.android.LocalBroadCastUtil
import dagger.android.AndroidInjection
import javax.inject.Inject

class DeleteWalletService @JvmOverloads constructor(name: String = TAG) : IntentService(name) {
    @Inject
    internal lateinit var cnWalletManager: CNWalletManager
    @Inject
    internal lateinit var pushNotificationDeviceManager: PushNotificationDeviceManager
    @Inject
    internal lateinit var pushNotificationEndpointManager: PushNotificationEndpointManager
    @Inject
    internal lateinit var apiClient: SignedCoinKeeperApiClient
    @Inject
    internal lateinit var analytics: Analytics
    @Inject
    internal lateinit var localBroadCastUtil: LocalBroadCastUtil
    @Inject
    internal lateinit var application: CoinKeeperApplication
    @Inject
    internal lateinit var syncWalletManager: SyncWalletManager
    @Inject
    internal lateinit var myTwitterProfile: MyTwitterProfile

    override fun onCreate() {
        AndroidInjection.inject(this)
        super.onCreate()
        localBroadCastUtil = LocalBroadCastUtil(this)
    }

    public override fun onHandleIntent(intent: Intent?) {
        syncWalletManager.cancelAllScheduledSync()
        if (pushNotificationEndpointManager.hasEndpoint()) {
            pushNotificationEndpointManager.unRegister()
            pushNotificationEndpointManager.removeEndpoint()
        }
        pushNotificationDeviceManager.removeCNDevice()
        apiClient.resetWallet()
        cnWalletManager.deleteWallet()
        resetAnalyticsProperties()
        myTwitterProfile.clear()
        localBroadCastUtil.sendBroadcast(DropbitIntents.ACTION_ON_WALLET_DELETED)
    }

    private fun resetAnalyticsProperties() {
        analytics.trackEvent(Analytics.EVENT_WALLET_DELETE)
        analytics.setUserProperty(Analytics.PROPERTY_HAS_WALLET, false)
        analytics.setUserProperty(Analytics.PROPERTY_PHONE_VERIFIED, false)
        analytics.setUserProperty(Analytics.PROPERTY_HAS_WALLET_BACKUP, false)
        analytics.setUserProperty(Analytics.PROPERTY_HAS_BTC_BALANCE, false)
        analytics.setUserProperty(Analytics.PROPERTY_HAS_DROPBIT_ME_ENABLED, false)
        analytics.setUserProperty(Analytics.PROPERTY_TWITTER_VERIFIED, false)
        analytics.flush()
    }

    companion object {
        val TAG = DeleteWalletService::class.java.name
    }

}
